### DISCUSSION TITLE
As a gamer, I want to have multiple playthroughts with different mod lists

### DISCUSSION THREAD

**Discussion by rockerbacon** [rockerbacon/ice-cream-machine-core#6] [Feature]
### How I do it today
Today, I have to keep multiple copies of my games. When I want to start a new playthrough, I create a copy of the unmodified game and install the mods I want to this copy.
For my save files, I do roughly the same, moving existing save files to a backup folder and creating a new empty directory for the new saves.
I then juggle these different copies and backups whenever I want to switch from one playthrough to another.

### How I would like to do it
I would like to have a tool to easily switch between playthroughs and have this tool change all the mods and save files automatically.

**Comment by rockerbacon** [comment]
### Feature suggestion
We will provide an interface through which clients can list and create profiles and also another which allows clients to select a profile to be the currently active profile.

#### Profile creation
We will always have an initial profile `Vanilla`, which contains the original game state and cannot be modified. The creation of a new profile will be the copy of this `Vanilla` profile.
Any new profile is initiated without any save files, this extends to the initial creation of the `Vanilla` profile.

#### Profile listing
Simple listing of all created profiles, including the `Vanilla` one.

#### Active profile selection
Whenever a client picks a new profile to be the active one, the system will symlink the selected profile to the location the game launcher (Steam, GOG, etc) will use to start the game.
The system should also symlink the save files.

**Reply by nlaranjeira01** [reply]
Will there be a copy of the entire game for each profile? Wouldn't it consume too much disk?

I was thinking of just replacing the files for each mod on the original game (while backing up the original game files) every time a profile is picked up.

**Reply by rockerbacon** [reply]
Indeed we must be careful here. These copies should be made using either hard-links or symlinks, so that no additional space will be used for files shared with the `Vanilla` profile.

**Comment by rockerbacon** [comment]
A recent issue arose in the gaming market that I feel we should start addressing from day 0. CAPCOM wants to kill modding and Bethesda is disregarding the open modding community in favour of its own mod marketplace.

Both companies are operating in the same way, releasing game updates that break modlists. Game clients then run their automatic updates and players are left with a broken modlist.

I believe we should ensure some level of profile immutability, that is, ice-cream-machine is allowed to modify game and mod files but no other process is. On Linux, I know we can achieve this by having a separate "icm" user (or any other name) and having this user own all game and mod files - this plus API authentication should be enough to keep users under control of their game states. I believe something similar should be achievable on MacOS but I have no idea on how to do it on Windows.

**Reply by nlaranjeira01** [reply]
On windows we can also create another user, give it ownership of folders and files, and remove permissions from other users as well.
We just need to be careful when dealing with permissions on Windows because it's possible to remove all Access Control Entries from the ACL, and reverting these cases is not trivial.

Also, when creating a user for ICM, we need a way to hide it from the user's login screen. I think it is possible to do it editing some values on the Registry Editor, but I am not fond of this solution and I also don't know how to do it programatically yet.

**Reply by nlaranjeira01** [reply]
Some commands.

Add user: `net user <username> <password> /add`. not sure if password can be blank.
To give ownership of a folder and files:
1. Grant full permissions to administrator and system users so that they are not inherited: `icacls <file or folder> /grant <adm or system>:F`
2. Set owner of the file or folder to ICM user: `icacls <file or folder> /setowner <icm user>`
3. Remove inherited ACEs from file or folder: `icacls <file or folder> /inheritancelevel:r`

**Reply by rockerbacon** [reply]
Maybe using Managed Service Accounts or Virtual Accounts would be preferable over regular users? Could help with hiding the user from the login screen
